sys/vm_map.c

```diff
@@ -132,10 +137,57 @@ void vm_map_protect(vm_map_t *map, vm_addr_t start, vm_addr_t end,
                     vm_prot_t prot) {
 }
 
+int vm_map_findspace(vm_map_t *map, vm_addr_t start, size_t length,
+                     vm_addr_t /*out*/ *addr) {
+
+  assert(is_aligned(length, PAGESIZE));
+  /* Bounds check */
+  if (start < map->pmap->start)
+    start = map->pmap->start;
+  if (start + length >= map->pmap->end)
+    return -ENOMEM;
+
+  if (TAILQ_EMPTY(&map->list)) {
+    /* Entire space free. */
+    *addr = start;
+    return 0;
+  }
+
+  vm_map_entry_t *first = TAILQ_FIRST(&map->list);
+  if (start + length <= first->start) {
+    /* If there is enought space before the first entry in the map, use it. */
+    *addr = start;
+    return 0;
+  }
+
+  /* Browse available gaps. */
+  vm_map_entry_t *it;
+  TAILQ_FOREACH (it, &map->list, map_list) {
+    vm_map_entry_t *next = TAILQ_NEXT(it, map_list);
+    if (!next)
+      continue;
+    if (next->start - it->end >= length) {
+      /* We will fit into this gap. */
+      *addr = it->end;
+      return 0;
+    }
+  }
+
+  /* Finally, check for free space after end. */
+  vm_map_entry_t *last = TAILQ_LAST(&map->list, vm_map_list);
+  if (map->pmap->end - last->end > length) {
```

Comment:
Again I can see a problem with the last page... `(map->pmap->end - last->end >= length)` ?